mirdata/web_downloader.py

```diff
@@ -0,0 +1,201 @@
+# -*- coding: utf-8 -*-
+"""functions for downloading from the web
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from collections import namedtuple
+import os
+import tarfile
+import shutil
+import zipfile
+import requests
+from requests.exceptions import HTTPError
+from tqdm import tqdm
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path  # python 2 backport
+
+from mirdata.utils import md5
+
+
+RemoteFileMetadata = namedtuple('RemoteFileMetadata', ['filename', 'url', 'checksum'])
+
+
+def downloader(save_dir, zip_downloads=None, tar_downloads=None,
+               file_downloads=None, info_message=None, force_overwrite=False):
+    """Download data to `save_dir` and optionally print a message.
+
+    Args:
+        save_dir (str):
+            The directory to download the data
+        zip_downloads (list or None):
+            A list of RemoteFileMetadata tuples of data in zip format.
+            If None, there is no zip data to download
+        tar_downloads (list or None):
+            A list of RemoteFileMetadata tuples of data in tar format.
+            If None, there is no tar data to download
+        file_downloads (list or None):
+            A list of  RemoteFileMetadata tuples of uncompressed data.
+            If None, there is no uncompressed data to download.
+        info_message (str or None):
+            A string of info to print when this function is called.
+            If None, no string is printed.
+        force_overwrite (bool):
+            If True, existing files are overwritten by the downloaded files.
+
+    """
+    Path(save_dir).mkdir(exist_ok=True)
```

Comment:
Is there a chance that we will ever have zip, tar and file downloads all in the same dataset? If not, it'd probably be easier to split this up into separate functions depending on the download type and call those functions separately.

Reply:
seldom both zip and tar, but often zip + file and tar + file. 

For one simple reason, I'd like to keep one mega function - we're going to test everything in `web_downloader`, we're not going to test each `loader.download()`. If we can have each `loader.download()` be a very thin wrapper on top of `web_downloader.downloader()`, the lack of test coverage isn't as bad as if each `loader.download()` has it's own logic.